src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -0,0 +1,307 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+
+namespace Microsoft.CommonLanguageServerProtocol.Framework;
+
+public abstract class AbstractLanguageServer<TRequestContext> : ILifeCycleManager, IAsyncDisposable
+{
+    private readonly JsonRpc _jsonRpc;
+    private IRequestExecutionQueue<TRequestContext>? _queue;
+    protected readonly ILspLogger _logger;
+    private ILspServices? _lspServices;
+
+    public bool IsInitialized { get; private set; }
+
+    // Fields used during shutdown.
+    private bool _shuttingDown;
+
+    public bool HasShutdownStarted => _shuttingDown;
+
+    protected AbstractLanguageServer(
+        JsonRpc jsonRpc,
+        ILspLogger logger)
+    {
+        _logger = logger;
+
+        _jsonRpc = jsonRpc;
+        _jsonRpc.AddLocalRpcTarget(this);
+        _jsonRpc.Disconnected += JsonRpc_Disconnected;
+    }
+
+    /// <summary>
+    /// This spins up the LSP and should be called when you're ready for your server to start receiving requests.
+    /// </summary>
+    public virtual Task InitializeAsync()
```

Comment:
So my best understanding here is that you're correct that this doesn't need to be async, but it DOES need to not be in the abstract constructor. Since this spins up the `RequestExecutionQueue` it needs the `ServiceProvider` to construct it. The `ServiceProvider` can be (and in the case of Roslyn is) provided in the constructor of the implementing class, which has not yet been run at the bottom of the abstract constructor. Will react accordingly.